cpp/src/arrow/json/reader.cc

```diff
@@ -42,132 +42,442 @@ namespace arrow {
-      }
-      RETURN_NOT_OK(parser->Parse(straddling));
+// Adds async-reentrancy to `source` by submitting tasks to a single-threaded executor
+// (FIFO order) - ensuring, at most, one future is pending at a time
+template <typename T>
+Result<AsyncGenerator<T>> MakeReentrantGenerator(AsyncGenerator<T> source) {
+  struct State {
+    AsyncGenerator<T> source;
+    std::shared_ptr<ThreadPool> thread_pool;
+  } state{std::move(source), nullptr};
+  ARROW_ASSIGN_OR_RAISE(state.thread_pool, ThreadPool::Make(1));
+
+  return [state = std::make_shared<State>(std::move(state))]() -> Future<T> {
+    auto maybe_future =
+        state->thread_pool->Submit([state] { return state->source().result(); });
+    return DeferNotOk(std::move(maybe_future));
+  };
+}
+
+class StreamingReaderImpl : public StreamingReader {
+ public:
+  StreamingReaderImpl(DecodedBlock first_block, AsyncGenerator<DecodedBlock> source,
+                      const std::shared_ptr<DecodeContext>& context, int max_readahead)
+      : first_block_(std::move(first_block)),
+        schema_(first_block_->record_batch->schema()),
+        bytes_processed_(std::make_shared<std::atomic<int64_t>>(0)) {
+    // Set the final schema for future invocations of the source generator
+    context->SetStrictSchema(schema_);
+    if (max_readahead > 0) {
+      source = MakeReadaheadGenerator(std::move(source), max_readahead);
     }
+    generator_ = MakeMappedGenerator(
+        std::move(source), [counter = bytes_processed_](const DecodedBlock& out) {
+          counter->fetch_add(out.num_bytes);
+          return out.record_batch;
+        });
+  }
 
-    if (whole->size() != 0) {
-      RETURN_NOT_OK(parser->Parse(whole));
+  static Future<std::shared_ptr<StreamingReaderImpl>> MakeAsync(
+      std::shared_ptr<DecodeContext> context, std::shared_ptr<io::InputStream> stream,
+      io::IOContext io_context, Executor* cpu_executor, const ReadOptions& read_options) {
+    if (!cpu_executor) {
+      cpu_executor = GetCpuThreadPool();
     }
 
-    std::shared_ptr<Array> parsed;
-    RETURN_NOT_OK(parser->Finish(&parsed));
-    builder_->Insert(block_index, field("", parsed->type()), parsed);
-    return Status::OK();
+    ARROW_ASSIGN_OR_RAISE(
+        auto buffer_it,
+        io::MakeInputStreamIterator(std::move(stream), read_options.block_size));
+    ARROW_ASSIGN_OR_RAISE(
+        auto buffer_gen,
+        MakeBackgroundGenerator(std::move(buffer_it), io_context.executor()));
+    buffer_gen = MakeTransferredGenerator(std::move(buffer_gen), cpu_executor);
+
+    auto chunking_gen = MakeChunkingGenerator(std::move(buffer_gen),
+                                              MakeChunker(context->parse_options()));
+    ARROW_ASSIGN_OR_RAISE(chunking_gen, MakeReentrantGenerator(std::move(chunking_gen)));
+
+    auto decoding_gen = MakeDecodingGenerator(std::move(chunking_gen),
+                                              DecodingOperator(context), cpu_executor);
```

Comment:
I didn't notice this on the previous review, but the CPU executor should ideally not be used if `use_threads` is false, i.e. every CPU-heavy task should happen on the thread calling `ReadNext()` (and, I suppose, `ReadNextAsync()` would then be blocking...).